### PR TITLE
Add URDF CLI converter

### PR DESCRIPTION
**Pull request by lukicdarkoo**
Currently, in ROS we are struggling to get URDF of a PROTO file. This PR should add a CLI verb that allows conversion from Webots PROTO to URDF and other supported formats (like `wbo` and `wrl`). The command should look like this:
```bash
webots convert -p extensionSlot="Box {}" ${WEBOTS_HOME}/projects/robots/adept/pioneer3/protos/Pioneer3dx.proto exported.urdf
```

That way we will be able to use Webots PROTOs just like `xacro`.

**Documentation**
https://cyberbotics.com/doc/guide/starting-webots?version=feature-cli-converter

**Tasks**
  - [x] Add basic exporter
  - [x] Fix URDF export
  - [x] Pass arguments
  - [x] Add field parsers
  - [x] Document
  - [x] Changelog

**Performance**
We anticipated low performance as Webots has to load a lot of shared libraries. 

The following log shows a converting time of the Pioneer 3-DX PROTO to VRML (starting Webots, parsing, exporting, writing to the file):
```
real    0m0.456s
user    0m0.308s
sys     0m0.137s
```
I believe that the result (just under 0.5s) is very good, probably more efficient than a standalone script written in Python. 

**Important notes**

This is the minimal code for creating a PROTO instance:
```c++
// First, we populate the global variables (paths) in the `WbProtoList`
// `WbProtoList` will be later used to find sub-PROTOs
new WbProtoList(QFileInfo(inputFile).absoluteDir().path());

// Create a PROTO model
// When the `InstantiateMode` is turned off then the parser creates only `WbNode`, not derived classes (e.g. WbGroup)
// The `InstantiateMode` has to be turned off here as `WbProtoList::readModel` will not work otherwise
// The `worldPath` parameter is irrelevant, it is passed all over the parser just to be handed over to the Lua context
WbNode::setInstantiateMode(false);
WbProtoModel *model = WbProtoList::current()->readModel(inputFile, "");

// We need create an array of `WbField` objects for the PROTO instance
// We can use default values from the model or override them
QVector<WbField *> fields;
for (WbFieldModel *model : model->fieldModels())
  fields.append(new WbField(model));

// For exporting the URDF we need `WbNode` derived objects so we have to turn on the `InstantiateMode` 
WbNode::setInstantiateMode(true);
WbNode *node = WbNode::regenerateProtoInstanceFromParameters(model, fields, false, "");
```

### CODE REVIEW

**Comment by lukicdarkoo** [comment]
I am not sure about the naming and a general structure for the subcommand as it is the first one. Let me know what do you think.

File: docs/reference/changelog-r2021.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ Released on June, Xth, 2021.
     - Exposed `stopERP` and `stopCFM` parameters in [HingeJointParameters](hingejointparameters.md) that define the local `ERP` and `CFM` used by joint limits.
     - Made the [TIAGo](../guide/tiago-base.md) robot more sturdy by increasing the `suspensionSpringConstant` value ([#2876](https://github.com/cyberbotics/webots/pull/2876)).
     - Altered the collision detection logic for [Robot.selfCollision](robot.md) to ignore chains of joints if the intermediary joints all share the same `anchor` point ([#2868](https://github.com/cyberbotics/webots/pull/2868)).
+    - Allow the URDF exportation from the command line ([#2885](https://github.com/cyberbotics/webots/pull/2885)).
```
Comment (omichel) [inline docs/reference/changelog-r2021.md:32]:
```suggestion
    - Added conversion from PROTO to URDF from the webots command line ([#2885](https://github.com/cyberbotics/webots/pull/2885)).
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 2]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
...
+    WbField *field = new WbField(model);
+    if (userParameters.contains(field->name())) {
+      WbTokenizer tokenizer;
+      tokenizer.tokenizeString(userParameters[field->name()]);
+      field->readValue(&tokenizer, "");
+    }
+
+    cout << tr("  field %1 [%2] = %3")
+              .arg(field->name())
+              .arg(field->stringType())
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:110]: This is more consistent for users:
```suggestion
              .arg(field->value()->vrmlTypeName())
```
and we could remove the stringType() method.

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 3]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
...
 
+void WbSingleTaskApplication::convertProto() const {
+  QCommandLineParser cliParser;
+  cliParser.setApplicationDescription("Convert a PROTO file to URDF, WBO, or WRL file");
+  cliParser.addHelpOption();
+  cliParser.addPositionalArgument("input", "Path to the input PROTO file.");
+  cliParser.addPositionalArgument("output", "Path to the output URDF, WBO, or WRL file.");
+  cliParser.addOption(QCommandLineOption("p", "Override default PROTO parameters.", "parameter=value"));
+  cliParser.process(mTaskArguments);
+  QStringList positionalArguments = cliParser.positionalArguments();
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:69]:
```suggestion
  const QStringList positionalArguments = cliParser.positionalArguments();
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 4]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
...
+              .arg(field->value()->toString())
+              .toUtf8()
+              .constData()
+         << endl;
+    fields.append(field);
+  }
+
+  // Generate a node structure
+  WbNode::setInstantiateMode(true);
+  WbNode *node = WbNode::regenerateProtoInstanceFromParameters(model, fields, false, "");
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:120]:
```suggestion
  const WbNode *node = WbNode::regenerateProtoInstanceFromParameters(model, fields, false, "");
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 5]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
...
+      cout << tr("A parameter is not properly formated!\n").toUtf8().constData();
+      cliParser.showHelp(1);
+    }
+    userParameters[pair[0]] = pair[1].replace(QRegExp("^\"*"), "").replace(QRegExp("\"*$"), "");
+  }
+
+  // Parse PROTO
+  new WbProtoList(QFileInfo(inputFile).absoluteDir().path());
+  WbNode::setInstantiateMode(false);
+  WbProtoModel *model = WbProtoList::current()->readModel(inputFile, "");
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:95]:
```suggestion
  const WbProtoModel *model = WbProtoList::current()->readModel(inputFile, "");
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 6]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
   else if (mTask == WbGuiApplication::VERSION)
     cout << tr("Webots version: %1").arg(WbApplicationInfo::version().toString(true, false, true)).toUtf8().constData() << endl;
   else if (mTask == WbGuiApplication::UPDATE_PROTO_CACHE)
-    updateProtoCacheFiles(mTaskArgument);
+    updateProtoCacheFiles();
   else if (mTask == WbGuiApplication::UPDATE_WORLD)
     WbWorld::instance()->save();
+  else if (mTask == WbGuiApplication::CONVERT) {
+    convertProto();
+  }
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:56]:
```suggestion
  else if (mTask == WbGuiApplication::CONVERT)
    convertProto();
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 7]
```diff
@@ -44,13 +48,91 @@ void WbSingleTaskApplication::run() {
...
+  QString outputFile = positionalArguments[1];
+  if (QDir::isRelativePath(inputFile))
+    inputFile = mStartupPath + '/' + inputFile;
+  if (QDir::isRelativePath(outputFile))
+    outputFile = mStartupPath + '/' + outputFile;
+
+  // Get user parameters strings
+  QMap<QString, QString> userParameters;
+  for (QString param : cliParser.values("p")) {
+    QStringList pair = param.split("=");
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:84]:
```suggestion
    const QStringList pair = param.split("=");
```

**Review by omichel — CHANGES_REQUESTED** [review]
I believe some const could be added here and there.

File: src/webots/gui/WbSingleTaskApplication.cpp (hunk 5 above)
Comment (lukicdarkoo) [reply src/webots/gui/WbSingleTaskApplication.cpp:94]: `regenerateProtoInstanceFromParameters` mutates the `model`

File: src/webots/gui/WbSingleTaskApplication.cpp (hunk 3 above)
Comment (lukicdarkoo) [reply src/webots/gui/WbSingleTaskApplication.cpp]: This cannot be const as the object is later mutated with `pair[1].replace(...`

**Comment by lukicdarkoo** [comment]
Actually, I think it is better to print to stdout

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 8]
```diff
@@ -44,13 +48,111 @@ void WbSingleTaskApplication::run() {
...
+  node->write(writer);
+  writer.writeFooter();
+
+  // Output the content
+  if (toStdout)
+    cout << output.toUtf8().toStdString() << endl;
+  else {
+    QFile file(outputFile);
+    if (!file.open(QIODevice::WriteOnly)) {
+      cout << tr("Cannot open the file!\n").toUtf8().constData();
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:145]:
```suggestion
      cerr << tr("Cannot open the file!\n").toUtf8().constData();
```

File: src/webots/gui/WbSingleTaskApplication.cpp [hunk 9]
```diff
@@ -44,13 +48,111 @@ void WbSingleTaskApplication::run() {
...
+    inputFile = mStartupPath + '/' + inputFile;
+  if (!toStdout && QDir::isRelativePath(outputFile))
+    outputFile = mStartupPath + '/' + outputFile;
+
+  // Get user parameters strings
+  QMap<QString, QString> userParameters;
+  for (QString param : cliParser.values("p")) {
+    QStringList pair = param.split("=");
+    if (pair.size() != 2) {
+      cout << tr("A parameter is not properly formated!\n").toUtf8().constData();
```
Comment (omichel) [inline src/webots/gui/WbSingleTaskApplication.cpp:94]:
```suggestion
      cerr << tr("A parameter is not properly formated!\n").toUtf8().constData();
```

**Review by omichel — COMMENTED** [review]
I believe that error messages should be sent to stderr instead of stdout.

**Review by omichel — DISMISSED** [review]
Thank you.

**Review by omichel — APPROVED** [review]

**Pull request merged**